ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,801 @@
+	if err != nil {
+		return "", "", err
+	}
+
+	registerRequest.TotalResources = resources
+
+	registerRequest.ClientToken = &registrationToken
+	resp, err := client.standardClient.RegisterContainerInstance(&registerRequest)
+	if err != nil {
+		logger.Error("Unable to register as a container instance with ECS", logger.Fields{
+			field.Error: err,
+		})
+		return "", "", err
+	}
+
+	var availabilityzone = ""
+	if resp != nil {
+		for _, attr := range resp.ContainerInstance.Attributes {
+			if aws.StringValue(attr.Name) == azAttrName {
+				availabilityzone = aws.StringValue(attr.Value)
+				break
+			}
+		}
+	}
+
+	logger.Info("Registered container instance with cluster!")
+	err = validateRegisteredAttributes(registerRequest.Attributes, resp.ContainerInstance.Attributes)
+	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), availabilityzone, err
+}
+
+func (client *ecsClient) setInstanceIdentity(
+	registerRequest ecsmodel.RegisterContainerInstanceInput) ecsmodel.RegisterContainerInstanceInput {
+	instanceIdentityDoc := ""
+	instanceIdentitySignature := ""
+
+	if client.configAccessor.NoInstanceIdentityDocument() {
+		logger.Info("Fetching Instance ID Document has been disabled")
+		registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+		registerRequest.InstanceIdentityDocumentSignature = &instanceIdentitySignature
+		return registerRequest
+	}
+
+	iidRetrieved := true
+	backoff := retry.NewExponentialBackoff(setInstanceIdRetryBackoffMin, setInstanceIdRetryBackoffMax,
+		setInstanceIdRetryBackoffJitter, setInstanceIdRetryBackoffMultiple)
+	ctx, cancel := context.WithTimeout(context.Background(), setInstanceIdRetryTimeOut)
+	defer cancel()
+	err := retry.RetryWithBackoffCtx(ctx, backoff, func() error {
+		var attemptErr error
+		logger.Debug("Attempting to get Instance Identity Document")
+		instanceIdentityDoc, attemptErr = client.ec2metadata.GetDynamicData(ec2.InstanceIdentityDocumentResource)
+		if attemptErr != nil {
+			logger.Debug("Unable to get instance identity document, retrying", logger.Fields{
+				field.Error: attemptErr,
+			})
+			// Force credentials to expire in case they are stale but not expired.
+			client.credentialsProvider.Expire()
+			client.credentialsProvider = instancecreds.GetCredentials(client.configAccessor.External())
+			return apierrors.NewRetriableError(apierrors.NewRetriable(true), attemptErr)
+		}
+		logger.Debug("Successfully retrieved Instance Identity Document")
+		return nil
+	})
+	if err != nil {
+		logger.Error("Unable to get instance identity document", logger.Fields{
+			field.Error: err,
+		})
+		iidRetrieved = false
+	}
+	registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+
+	if iidRetrieved {
+		instanceIdentitySignature, err = client.ec2metadata.
+			GetDynamicData(ec2.InstanceIdentityDocumentSignatureResource)
+		if err != nil {
+			logger.Error("Unable to get instance identity signature", logger.Fields{
+				field.Error: err,
+			})
+		}
+	}
```

Comment:
Why have we not opted for exponential backoff in this api call?

Reply:
We have not ran into any transient failures upstream with the call on L298 to motivate adding a retry mechanism for this call to get instance identity signature on L297-L298.

For the call to get instance identity document on L275 we ran into multiple issues of customers facing transient failures upstream, which was what motivated https://github.com/aws/amazon-ecs-agent/pull/3795.

Reply:
Discussed offline, despite not having run into any transient failures upstream with the call to get instance identity signature, we will proactively add the same retry mechanism for this call in a follow up pull request to allow Agent to be more resilient in the event that we do run into any transient failures upstream with this call in the future.